✨ feat  : 적성 검사 이어하기 및 AI 대화형 적성 검사 재설계

sooyeoneo opened the pull request:
## ☀️ 작업 목적

- 적성 검사의 **부분 저장 및 이어하기 기능**을 제공하여 사용자 경험(UX)을 개선
- 고정형 질문 Pool을 제거하고 **AI 대화형 기반**으로 재설계
- Redis 세션/질문-답변 캐시 관리 강화로 **데이터 무결성과 연속성 보장**
- 타입 안정성(Enum), 서비스 책임 분리, 세션 보안 강화

---

## 📄 주요 작업 내역

### 1. Entity & Repository 개선

- **AptitudeTestHistory**
    - `aptitudeType: String` → `AptitudeType analyzedAptitudeType (Enum)`
    - `isPartialSave`, `lastCompletedStep` 필드 추가
    - 세션 ID, 신뢰도 점수, 완료 여부 관리
- **UserAptitude**
    - `lastPartialStep`, `partialSessionId` 필드 추가
    - `updatePartialProgress()`, `clearPartialProgress()` 메소드
- **AptitudeTestHistoryRepository**
    - `findTopByMemberAndIsCompletedFalseOrderByCreatedAtDesc()` 미완료 세션 조회
    - `findMaxStepBySessionId()`, `deleteByMemberAndIsCompletedFalse()` 추가
    - `findByMemberAndSessionIdOrderByStepAsc()` → Member 파라미터 필수 (권한 검증 강화)

---

### 2. Service/로직 리팩토링

- **AptitudeService**
    - `savePartialResult()`: 모든 단계마다 자동 부분 저장
    - `resumeTest()`: 중단된 세션 이어하기 (Redis 데이터 복구 포함)
    - `submitAnswer()`: TTL 갱신 및 세션 유효성 검증
    - `convertToDto()`, `parseAptitudeType()` 내부 로직 정리
- **AptitudeTestHistoryService**
    - 테스트 이력 저장/조회/삭제 전담
    - 세션 관리 및 완료 처리 분리 → SRP 준수
- **AptitudeCacheService**
    - `saveSessionId()`, `getSessionId()`, `refreshSession()`
    - `saveQuestion()`, `getQuestion()`, `saveFollowUpQuestion()` → 대화 맥락 관리

---

### 3. Controller & API

- **AptitudeController**
    - `GET /ai/aptitude/resume`: 이어하기 API 추가
    - `GET /ai/aptitude/history`: 완료 이력 조회
    - `GET /ai/aptitude/history/session`: 세션별 이력 조회
- **DTO**
    - `AptitudeResumeResponseDto`: 이어하기 응답 DTO
    - `AiQuestionGenerationDto`: AI 질문 + 예시 답변 전달
    - `AiAnswerValidationDto`: 답변 유효성 검증 및 재질문 포함
    - DTO는 데이터만 담당, 로직은 Service로 이동

---

### 4. AI 적성 검사 대화형 리팩토링

- 고정 질문 Pool 제거 → **AI 동적 질문 생성**
- `buildConversationHistory()`: 대화 히스토리 구성
- 무효 답변 감지 및 follow-up 질문 로직 추가
- Temperature 0.7 적용 → 창의적 질문 생성
- Redis에 질문/답변 히스토리 저장 → 대화 연속성 유지
- Fallback 로직 강화 (5개 적성 타입별 일관 구조)

---

### 5. Time-to-Live 관리

- 세션 TTL **30분 → 60분 연장**
- 답변 제출 시 TTL 자동 갱신
- 방치된 세션 자동 만료 → 메모리 효율성 확보

---

### 6. Bug Fixes

- **NOT NULL 제약조건 오류**: `UserAptitude` Builder 시 임시 기본값 `NATURE` 설정
<img width="1792" height="1137" alt="UserAptitude 엔티티 생성시 NOT NULL 제약조건 오류 2025-09-15 18 44 09" src="https://github.com/user-attachments/assets/14c562a7-9a09-4cca-bda9-9afa94da3585" />
<img width="1792" height="1137" alt="UserAptitude 엔티티 생성시 NOT NULL 제약조건 오류 2025-09-15 18 45 54" src="https://github.com/user-attachments/assets/e6e25c64-324d-4bec-a1c1-7178b99c4447" />


- **무효 답변 NPE**: `aptitudeType null 체크` 추가, 점수 업데이트 생략, "무효 답변" 처리
<img width="1792" height="1137" alt="무효 답변 NullPointException 2025-09-15 20 22 32" src="https://github.com/user-attachments/assets/ef382ab4-756b-4d9d-a251-1d94dc0f3ad8" />
<img width="1792" height="1137" alt="무효 답변 NullPointException 2025-09-15 20 31 11" src="https://github.com/user-attachments/assets/7e21fbe8-3f36-44bb-a24d-0a5d5efd9aa9" />


---

### 7. API 테스트

- `ai-aptitude.http` 테스트 케이스 강화
    - 이어하기 API 정상 동작 확인
    - 무효 답변/복합 답변/대화 연속성 케이스 포함
- Redis 로그, HTTP 응답 캡쳐 첨부 (테스트 증빙)
<img width="1792" height="1137" alt="2-3 기술중심 답변 시나리오API 테스트 2025-09-15 20 18 55" src="https://github.com/user-attachments/assets/f19da0c9-c703-4183-b059-102070b9d05b" />
<img width="1792" height="1137" alt="2-3 기술중심 답변 시나리오API 테스트 console 2025-09-15 20 18 59 1" src="https://github.com/user-attachments/assets/267eb254-ca0b-4730-b74e-bc9207b2e905" />


---

## 📸 테스트 결과 - 추가 예정

- **이어하기 API** 정상 동작
<img width="1792" height="1137" alt="적성 테스트 Step5 2025-09-03 06 33 28" src="https://github.com/user-attachments/assets/93adb777-41da-42e8-a7b2-8ec345c0ddf3" />

- **세션별 이력 조회**
<img width="1792" height="1137" alt="적성 테스트 Step1 로그2025-09-03 06 32 21" src="https://github.com/user-attachments/assets/bf826878-4f34-4a41-ae84-74825f393852" />

<img width="1792" height="1137" alt="적성 테스트 Step5 2025-09-03 06 33 28" src="https://github.com/user-attachments/assets/05348583-25f6-4df6-8187-234d1f988642" />

- **Redis 세션 로그**
<img width="1789" height="1127" alt="Redis CLI MONITOR 테스트 완료 2025-09-03 06 40 43" src="https://github.com/user-attachments/assets/015c1e66-d120-4875-9ef0-1ac3dbcc01e1" />

<img width="1789" height="1127" alt="Redis CLI MONITOR 테스트 완료 2025-09-03 06 40 50" src="https://github.com/user-attachments/assets/f52e61f5-9279-4f66-b237-ae5170d7e7d0" />

---

## 🛠️ 참고 및 TODO

1. **Optional 적극 활용 (Repository 중심)**
    - 단일 Step 조회, 마지막 Step 조회에 Optional 도입 예정
    - `.orElseThrow()` 활용으로 NullPointerException 원천 차단
2. **대화 상태 무결성 보장**
    - `currentStep` 필드 추가 및 요청 Step과 DB Step 일치 검증
    - 비관적 락(`PESSIMISTIC_WRITE`)으로 동시성 제어 필요
3. **가독성과 유지보수성**
    - `@Transactional(readOnly = true)` 적용으로 성능 최적화
    - 매직넘버(총 Step=5) → 상수 추출로 의미 부여
4. **추가 개선**
    - GlobalExceptionHandler 적용
    - API 문서(Swagger) 보강
    - 테스트 커버리지 확장 (TTL 만료, Optional empty, 동시성 race condition)

---

**리뷰 및 피드백 환영합니다! 🤗**

---

**Resolves**: #64 #65 #67 #98 #151 #157 #164 #166 #170

Note: The `Co-authored-by: @jiyoon0000` tag reflects code-review feedback,
not direct code implementation or PR document input.

sooyeoneo committed:
✨ feat : 적성 검사 Entity 및 Enum 구현

- UserAptitude Entity 구현
  - 회원별 적성 정보 저장 (1:1 관계)
  - 테스트 횟수 추적 (최대 5회 제한)
  - BaseEntity 상속으로 생성/수정 시간 자동 관리

- AptitudeTestHistory Entity 구현
  - 적성 검사 이력 저장 (N:1 관계)
  - 질문-답변-AI응답 기록

- AptitudeType Enum 구현
  - 5가지 적성 타입 정의 (자연친화, 역사문화, 예술창작, 커뮤니티, 디지털기술)
  - 각 타입별 제목과 설명 포함

#65

sooyeoneo committed:
✨ feat : 적성 검사 Repository 구현

- UserAptitudeRepository
  - 회원별 적성 조회 (findByMember)
  - 적성 존재 여부 확인 (existsByMember)

- AptitudeTestHistoryRepository
  - 회원별 검사 이력 조회 (오름차순 정렬)
  - 회원별 이력 전체 삭제

#65

sooyeoneo committed:
✨ feat : 적성 검사 DTO 구현

- RequestDto (요청)
  - AptitudeAnswerRequestDto: 사용자 답변 제출

- ResponseDto (응답)
  - AptitudeQuestionResponseDto: AI 질문 정보를 사용자에게 답변 응답
  - AptitudeTestStartResponseDto: 검사 시작 응답
  - AptitudeTextProgressResponseDto: 진행 상태 응답
  - AptitudeTestResultResponseDto: 최종 결과 응답
  - CanRetakeTestResponseDto: 재검사 가능 여부 응답

* 모든 ResponseDto는 불변 객체로 구현 (생성자 사용)

#64 #65

sooyeoneo committed:
✨ feat : 적성 검사 Service 구현

- AptitudeAiService
  - 5개 질문 Pool 정적 관리
  - AI 프롬프트 기반 답변 분석
  - 적성 타입별 점수 계산 및 최종 적성 결정
  - BaseAiService 상속으로 재시도 및 예외처리

- AptitudeService
  - 검사 시작/답변 제출/결과 저장 플로우
  - Redis 기반 검사 진행 상태 관리 (24시간 TTL)
  - 최대 5회 재검사 제한 로직
  - 수동 적성 선택 기능
  - CustomException 기반 예외처리

* @Transactional 메서드별 적용

#64 #65 #66

sooyeoneo committed:
✨ feat : 적성 검사 Controller 구현

- API 엔드포인트
  - POST /api/ai/aptitude/test/start: 검사 시작
  - POST /api/ai/aptitude/test/answer: 답변 제출
  - POST /api/ai/aptitude/select: 수동 적성 선택
  - GET /api/ai/aptitude/can-retake: 재검사 가능 여부

- Spring Security 적용
  - @PreAuthorize(\"hasRole('USER')\") 일반회원 권한 체크
  - MemberDetails 통한 회원 정보 획득

- CommonResponseDto 표준 응답 형식 적용

#63 #66

sooyeoneo committed:
🔧 add : Redis 범용 aiAptitudeRedisTemplate 빈 추가

- RedisTemplate<String, String> 빈 추가
- @Qualifier("aiAptitudeRedisTemplate") 빈에 추가 구분자를 붙여 명시적으로 지정
- 적성 검사 진행 상태 관리용
- 기존 blacklist, emailVerified 템플릿과 별도 관리

#64

sooyeoneo committed:
✅ test : 적성 검사 HTTP 테스트 및 AI 프롬프트 추가

- ai-aptitude.http
  - 적성 검사 시작/답변/선택/재검사 API 테스트
  - Authorization Bearer 토큰 기반

- aptitude-test.txt
  - 적성 분석 AI 프롬프트 추가

#67

sooyeoneo committed:
💡 comment : AiTestController 주석 처리 (추후 삭제 예정)

sooyeoneo committed:
💡 comment : 필요 없는 주석 삭제

sooyeoneo committed:
🔥 remove : @PreAuthorize("hasRole('USER')") 제거

sooyeoneo committed:
🔧 add : NOT_FOUND_TEST_PROGRESS, TOO_MANY_REQUESTS ErrorCode에 추가

sooyeoneo committed:
🔧 add : 적성 검사 AI 응답 분석에 정규식 패턴 매칭 로직 추가

- analyzeHistories() 메서드에 Pattern/Matcher 기반 분석 로직 적용
- 정규식으로 모든 적성 타입(name + title) OR 연결 패턴 생성
- AI 응답 내 복수 적성 키워드 추출 및 점수 계산 지원
- 향후 HashMap 최적화를 위한 TODO 주석 추가

Refs: #65

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

# Conflicts:
#	src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeService.java
#	src/main/java/com/salayo/locallifebackend/global/config/RedisConfig.java
#	src/main/java/com/salayo/locallifebackend/global/error/ErrorCode.java

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

sooyeoneo committed:
🔧 add : @PreAuthorize("hasRole('USER')") 추가

- AptitudeController에 @PreAuthorize("hasRole('USER')") 일반 회원 권한을 다시 추가

#55

sooyeoneo committed:
🔧 add : @RequireArgsConstructor 삭제하고, 생성자 직접 주입

- AptitudeController에 @RequireArgsConstructor 삭제하고, 생성자 직접 주입(Constructor injection)으로 의존성을 직접 전달

#55

sooyeoneo committed:
🔥 remove : 사용하지 않는 import 문 제거

sooyeoneo committed:
♻️ refactor : 적성 검사 캐시 로직 분리 및 CacheKeyPrefix로 통합 관리

- AptitudeCacheService 생성하여 Redis 캐시 로직 캡슐화
  - ReviewCacheService와 동일한 아키텍처 패턴 적용
  - 테스트 진행 상태 저장/조회/삭제 메서드 구현

- CacheKeyPrefix에 적성 검사 관련 모든 상수 통합
  - APTITUDE_MAX_TEST_COUNT: 최대 테스트 횟수 (5회)
  - APTITUDE_TOTAL_QUESTIONS: 전체 질문 개수 (5개)
  - APTITUDE_TEST: Redis 키 프리픽스 ("aptitude:test:")
  - APTITUDE_TEST_TTL_HOURS: 캐시 TTL (24시간)

- AptitudeService 리팩토링
  - 캐시 로직을 AptitudeCacheService로 위임
  - CacheKeyPrefix 상수 참조로 중앙 집중식 관리로 유지보수성 향상
  - AptitudeService와 AptitudeCacheService에서 CacheKeyPrefix 참조

#55 #65

sooyeoneo committed:
💡 comment : 주석 수정

sooyeoneo committed:
🐛 fix : 적성 검사 횟수 제한 에러코드 수정

- APTITUDE_TEST_LIMIT_EXCEEDED 에러코드 추가 (400 BAD_REQUEST)
- TOO_MANY_REQUESTS(429)는 Rate Limiting 전용으로 복원
- 비즈니스 로직 제약사항에 맞는 적절한 상태 코드 사용

#55 #65 #67

sooyeoneo committed:
📝 docs : 적성 검사 API Swagger 문서화 추가

- Controller 레벨
  - @Tag 추가: "Aptitude" 그룹화 및 설명
  - @Operation: 각 API 엔드포인트별 요약 및 상세 설명
  - @ApiResponses: HTTP 상태 코드별 응답 설명 (200, 201, 400, 401, 404)
  - @Parameter: 요청 파라미터 설명 및 예시값

- DTO 레벨
  - 모든 Request/Response DTO에 @Schema 추가
  - 필드별 설명, 예시값, 최소/최대값, 필수 여부 명시
  - AptitudeQuestionResponseDto: 질문 정보 및 예시 답변
  - AptitudeAnswerRequestDto: 답변 요청 필드 제약사항
  - CanRetakeTestResponseDto: 재검사 관련 정보

- Enum 레벨
  - AptitudeType에 @Schema 추가
  - 각 적성 타입별 상세 설명 포함

Swagger UI에서 적성 검사 API의 상세 문서 확인 가능

#55 #64 #65 #66

sooyeoneo committed:
🔧 add : @Valid 추가

sooyeoneo committed:
✏️ typos : 코드 줄 정리

sooyeoneo committed:
🔧 add : ResponseDto에 @Builder를 생성자 레벨에 추가

sooyeoneo committed:
♻️ refactor : 적성 검사 횟수 제한 로직 개선 - 온보딩/마이페이지 구분

- UserAptitude 엔티티 개선
  - mypageTestCount 필드 추가 (마이페이지 검사 횟수)
  - isOnboardingCompleted 필드 추가 (온보딩 완료 여부)
  - 온보딩/마이페이지별 업데이트 메서드 분리

- 적성 검사 제한 로직 변경
  - 온보딩 첫 검사는 횟수 제한 없음
  - 마이페이지에서만 최대 5회 제한 적용
  - 총 6회 검사 가능 (온보딩 1회 + 마이페이지 5회)

- API 문서 및 에러 메시지 업데이트
  - 온보딩과 마이페이지 검사 구분 명시
  - APTITUDE_TEST_LIMIT_EXCEEDED 메시지 수정

#55 #66

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

# Conflicts:
#	src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/controller/AptitudeController.java
#	src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/dto/CanRetakeTestResponseDto.java
#	src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeService.java
#	src/main/java/com/salayo/locallifebackend/global/error/ErrorCode.java

sooyeoneo committed:
♻️ refactor : AptitudeAiService 구조 개선 및 캐시 활용 강화

- AptitudeCacheService에 답변 저장 및 적성 점수 관리 기능 추가
  - saveAnswer(): 단계별 답변 Redis 저장
  - updateAptitudeScore(): 적성별 점수 누적 관리
  - getAllAptitudeScores(): 전체 점수 조회
  - deleteAllTestData(): 테스트 데이터 정리

- AptitudeAiService 단순화 및 개선
  - AI 호출 실패 시 fallback 키워드 분석
  - 로그 레벨 log.warn 설정

- AptitudeService Redis 활용 개선
  - AI 분석 결과를 Redis에 실시간 저장
  - DB 텍스트 파싱 대신 Redis 점수 직접 활용
  - 테스트 완료 후 자동 데이터 정리

#55 #65 #67 #151

sooyeoneo committed:
♻️ refactor : AI 적성 검사 테스트 환경 및 데이터 정리 로직 개선

- HTTP 테스트 파일 개선
  - 동적 변수 처리를 위한 응답 구조 추가(aptitudeQuestionResponseDto 경로 수정)
  - AI 적성 검사 전체 플로우, 개별 단계별, 수동 선택 테스트 섹션 구분
  - 콘솔 로그 추가로 테스트 진행 상황 추적 가능

- Redis 데이터 정리 로직 강화
  - deleteAllTestData() 메소드에 상세 로그 추가
  - 각 데이터 타입별 삭제 결과 확인 (진행상태, 답변, 점수)
  - 테스트 시작 시 기존 데이터 정리로 중복 방지

- 디버깅 로그 추가
  - Redis 데이터 정리 시작/완료 로그
  - 답변 저장 및 점수 업데이트 과정 추적

#55 #65 #67 #151

sooyeoneo committed:
🏗 chore : AI 적성 검사 세션 타임아웃 30분 설정

#55 #64

sooyeoneo committed:
✨ feat : AI 적성 분석 JSON 파싱 구조화 및 신뢰도 기반 점수 적용

- AiAptitudeAnalysisResponseDto 생성 (적성 타입, 신뢰도, 이유, 핵심요소)
- AiQuestionGenerationResponseDto 생성 (추후 동적 질문 생성용)
- AptitudeAiService에서 구조화된 객체 반환 및 JSON 응답 프롬프트 개선
- 신뢰도 점수(0.7 기준)에 따른 가중치 적용 (높음: 2점, 낮음: 1점)
- Fallback 분석도 동일한 객체 구조로 통일
- 테스트용 로그 추가 (JSON 파싱, 가중치 적용 확인)

#55 #64 #67 #151

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

sooyeoneo committed:
♻️ refactor : 코드 리뷰 반영 - AI 적성 분석 안정성 및 일관성 개선

[변경 내용]
1. AiAptitudeAnalysisResponseDto 개선
   - String[] keyFactors → List<String> keyFactors 변경
   - 방어적 복사(Defensive Copy) 적용으로 null 안전성 향상
   - 프로젝트 내 다른 DTO들과 컬렉션 사용 통일

2. AptitudeService 안전성 강화
   - toUpperCase() 적용으로 소문자 입력 대응
   - AI 응답의 대소문자 구분 없이 안전하게 처리

3. AptitudeAiService fallback 로직 개선
   - List.of() 사용으로 일관성 유지
   - Fallback 시에도 동일한 자료구조 사용

[개선 효과]
- 프로젝트 전체 일관성 확보
- null 및 대소문자 안전성 향상
- Spring/Jackson 표준 준수
- 유지보수성 개선

Resolves: #158
Co-authored-by: jiyoon0000 <rlawldbs9128@gmail.com>

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

sooyeoneo committed:
✨ feat : AptitudeTestHistory Entity 타입 안정성 개선

- String aptitudeType → AptitudeType analyzedAptitudeType (Enum 타입 통일)
- 세션 ID, 신뢰도 점수, 완료 여부 필드 추가
- UserAptitude와 동일한 Enum 타입 사용으로 일관성 확보

Resolves: #55 #151 #164

sooyeoneo committed:
♻️ refactor : AptitudeTestHistoryRepository 메소드 정리

- findByMemberOrderByStepAsc() 제거 (세션 기반으로 대체)
- 각 메소드 용도 명확히 하는 주석 추가
- markSessionAsCompleted() 쿼리 메소드로 일괄 완료 처리
- 추후 사용 예정 메소드 용도 명시

Resolves: #65 #164

sooyeoneo committed:
✨ feat : AptitudeService 세션 기반 이력 관리 구현

- generateSessionId(): APT-{userId}-{timestamp} 형식 세션 생성
- parseAptitudeType(): String → Enum 안전 변환 헬퍼 메소드
- convertToDto(): Entity → DTO 변환 로직 Service로 이동
- 테스트 완료 시 세션 일괄 완료 처리
- 세션별/완료된 이력 조회 메소드 구현

Resolves: #65 #151 #164

sooyeoneo committed:
✨ feat : AptitudeCacheService Redis 세션 관리 기능 추가

- saveSessionId(), getSessionId(): 세션 ID 저장/조회
- deleteAllTestData(): 세션 ID 포함 전체 데이터 정리
- 세션 기반 테스트 데이터 격리

Resolves: #65 #151 #164

sooyeoneo committed:
✨ feat : 적성 검사 이력 조회 API 및 DTO 개선

- GET /ai/aptitude/history: 완료된 이력 조회 API
- GET /ai/aptitude/history/session: 세션별 이력 조회 API
- AptitudeTestHistoryResponseDto from() 메소드 제거
- DTO 순수성 유지 (데이터만 담당, 로직은 Service)

Resolves: #65 #151 #164

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

sooyeoneo committed:
✅ test : ai-aptitude.http에 이력 조회 API 테스트 추가

[추가 내용]
- GET /ai/aptitude/history: 완료된 적성 검사 이력 조회
- GET /ai/aptitude/history/session: 특정 세션 ID별 이력 조회
- 세션 ID 예시 및 사용 가이드 콘솔 출력

[Test Flow]
1. 적성 테스트 완료 (5단계)
2. 완료된 이력 전체 조회
3. 반환된 세션 ID로 특정 세션 조회

Resolves: #65 #151 #164

sooyeoneo committed:
✨ feat : 적성 검사 세션 30분 타임아웃 구현

- Redis 세션 TTL 30분 설정
- 답변 제출 시마다 세션 갱신
- 30분 이상 미활동 시 자동 만료
- 세션 재생성 로직 및 로그 추가

[개선 효과]
- 사용자가 잠시 자리를 비워도 테스트 유지
- 방치된 세션 자동 정리로 메모리 효율화

Resolves: #65 #151 #164

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

sooyeoneo committed:
🐛 fix : AptitudeTestHistoryResponseDto의 aptitudeType의 타입 String → Enum 수정

Resolves: #164 #166

sooyeoneo committed:
🐛 fix : 유효하지 않은 ai-aptitude의 test session에 대한 명시적 예외 처리 추가

[변경 내용]
- 제공된 세션 ID로 이력 목록을 조회했을 때 비어 있는 경우, AptitudeService.getTestHistoryBySession 메서드가 SESSION_NOT_FOUND 예외를 발생시키도록 수정
- 세션 이력이 존재하는 경우에만 소유권 확인하도록 로직을 수정하여, 잠재적인 IndexOutOfBoundsException 발생 방지

[개선 효과]
- 유효하지 않거나 만료된 세션과 이력이 없는 유효한 세션을 구분하여, 존재하지 않는 세션에 접근하려는 사용자에게 더 명확하고 직관적인 오류 응답을 제공

Resolves: #164 #166

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

sooyeoneo committed:
🔧 add : 적성 검사 부분 저장을 위한 Entity 필드 추가

- AptitudeTestHistory: 부분 저장 관련 필드 추가
  - isPartialSave: 부분 저장 여부
  - lastCompletedStep: 마지막 완료 단계

- UserAptitude: 부분 저장 정보 필드 추가
  - lastPartialStep: 마지막 부분 저장 단계
  - partialSessionId: 부분 저장된 세션 ID
  - updatePartialProgress(), clearPartialProgress() 메소드 추가

Resolves: #151 #164 #166

sooyeoneo committed:
✨ feat : 적성 검사 이어하기 ResponseDto 구현

- AptitudeResumeResponseDto 생성
  - sessionId: 이어할 세션 ID
  - nextStep: 다음 진행 단계
  - totalSteps: 전체 단계 수
  - nextQuestion: 다음 질문 정보
  - completedSteps: 완료된 단계
  - message: 안내 메시지

Resolves: #151 #164 #166 #170

sooyeoneo committed:
🔧 add : 적성 검사 이어하기를 위한 Repository 메소드 추가

- AptitudeTestHistoryRepository 메소드 추가
  - findTopByMemberAndIsCompletedFalseOrderByCreatedAtDesc(): 미완료 세션 조회
  - findMaxStepBySessionId(): 세션의 최대 step 조회
  - deleteByMemberAndIsCompletedFalse(): 미완료 이력만 선택적 삭제

[개선 효과]
- 데이터 손실 방지
- 미완료 테스트만 정리하여 데이터 무결성 유지
- 데이터 손실 방지

Resolves: #151 #164 #166

sooyeoneo committed:
✨ feat : 적성 검사 부분 저장 및 이어하기 기능 구현

- AptitudeService 기능 추가
  - savePartialResult(): 모든 단계마다 자동 부분 저장
  - resumeTest(): 중단된 테스트 이어하기
  - Redis 데이터 복구 (세션, 점수, 답변)
  - 테스트 완료 시 부분 저장 정보 초기화

- ErrorCode 추가 (NO_INCOMPLETE_TEST)

[개선 효과]
- 모든 단계에서 자동 저장으로 부분 저장으로 데이터 보호
- 중단 후에도 이어서 진행 가능
- 사용자 경험 개선

Resolves: #151 #164 #166 #170

sooyeoneo committed:
🔧 add : 적성 검사 이어하기 API 및 테스트 추가

- AptitudeController
  - GET /ai/aptitude/resume: 중단된 테스트 이어하기 API

- ai-aptitude.http
  - 이어하기 테스트 케이스 추가
  - 응답 처리 스크립트 포함

Resolves: #151 #164 #166 #170

sooyeoneo committed:
♻️ refactor : AptitudeService 코드 정리

- analyzeHistories(): 불필요한 메소드 제거
- import 최적화
- getFirst(): 메소드 활용

Resolves: #151 #164

sooyeoneo committed:
✨ feat : 적성 검사 세션 60분 타임아웃 구현

- CacheKeyPrefix
  - SESSION_TIMEOUT_MINUTES 상수 설정 (기존 30분 → 60분)

- AptitudeCacheService
  - refreshSession(): 답변 제출 시 세션 TTL 갱신

- AptitudeService
  - submitAnswer(): 세션 유효성 확인 및 TTL 갱신 로직

[개선 효과]
- 사용자 편의성 향상 (60분간 유지)
- 방치된 데이터 자동 정리

Resolves: #151 #164 #166

sooyeoneo committed:
🐛 fix : 적성 검사 이력 조회 시 사용자 검증 추가

- AptitudeTestHistoryRepository
  - findBySessionIdOrderByStepAsc() → findByMemberAndSessionIdOrderByStepAsc()
  - Member 파라미터 추가로 권한 검증 강화

- AptitudeService
  - 이력 조회 메서드들 Member 파라미터 추가
  - 본인 세션만 조회 가능하도록 보안 강화

[개선 효과]
- 타 사용자 세션 접근 방지 및 본인 소유 세션만 조회 가능하도록 제한
- 무단 데이터 접근 방지로 데이터 보안성 향상

Resolves: #164 #166

sooyeoneo committed:
♻️ refactor : 적성 검사 이력 관련 로직을 별도 서비스로 분리

[변경 사항]
- AptitudeTestHistoryService 추가
  - 테스트 이력 저장/조회/삭제 로직 이관
  - 세션 관리 및 완료 처리 담당

- AptitudeService 리팩토링
  - 적성 검사 핵심 비즈니스 로직만 유지
  - 이력 관련 작업은 AptitudeTestHistoryService에 위임
  - Entity Builder 패턴으로 데이터 생성 후 전달

- AptitudeController 수정
  - TestHistoryService 주입 추가
  - 이력 조회 API는 HistoryService 직접 호출

[개선 효과]
- 단일 책임 원칙(SRP) 준수
- 서비스 간 책임 분리로 유지보수성 향상
- 코드 가독성 및 테스트 용이성 개선

Resolves: #164 #166

sooyeoneo committed:
🏗️ chore : 적성 검사 세션 60분 타임아웃

- application.properties
  - # AI Aptitude Test
    ai.aptitude.session.timeout=60 (기존 30분 → 60분)

Resolves: #151 #164 #166

sooyeoneo committed:
🔧 add : 적성 검사 캐시 서비스 기능 확장

[추가 기능]
- 질문/답변 저장 및 조회 메서드 추가
  - saveQuestion(), getQuestion()
  - getAnswer() 메서드 추가
  - saveFollowUpQuestion() 재질문 저장 기능

[목적]
- AI 대화형 적성 검사를 위한 대화 히스토리 관리
- 질문-답변 쌍을 Redis에 저장하여 맥락 유지

Resolves: #151 #164 #166 #170

sooyeoneo committed:
✨ feat : AI 적성 검사 내부 DTO 클래스 생성

[새로운 DTO]
- AiQuestionGenerationDto
  - AI가 생성한 질문과 5가지 예시 답변 전달
  - AI Service → Cache Service 데이터 전달용

- AiAnswerValidationDto
  - 사용자 답변 유효성 검증 결과 전달
  - 무효 답변 시 follow-up 질문 포함

[구조]
- Jackson 어노테이션으로 JSON 매핑
- 적성 타입별 예시 답변 관리

Resolves: #151 #157 #170

sooyeoneo committed:
♻️ refactor : AI 대화형 적성 검사로 전면 재설계

[주요 변경사항]
- 고정 질문 풀(QUESTION_POOL) 완전 제거
- AI가 대화 맥락 고려하여 동적 질문 생성
- 5가지 적성 타입별 예시 답변 자동 생성

[새로운 기능]
- buildConversationHistory(): 대화 히스토리 구성
- 무효 답변 감지 및 재질문 로직
- JSON 파싱 결과 상세 로깅
- Fallback 로직 강화 (5가지 적성별)

[개선사항]
- Temperature 0.7 활용한 창의적 질문
- 이전 Q&A 맥락을 AI에게 제공
- 하위 호환성 위한 오버로드 메서드

[Breaking Change]
- getNextQuestion() 시그니처 변경
- analyzeResponse() 파라미터 추가

Resolves: #64 #98 #170

sooyeoneo committed:
♻️ refactor : AI 대화형 적성 검사 통합

[변경사항]
- AI 질문 생성 시 memberId 전달
- 생성된 질문을 Redis에 저장
- 답변 분석 시 대화 맥락 포함

[수정 메서드]
- startTest(): AI 동적 질문 생성 적용
- submitAnswer(): 대화 맥락 전달
- resumeTest(): AI 질문 생성 적용

[목적]
- AI가 사용자별 맞춤 질문 생성
- 대화 연속성 보장

Resolves: #64 #98 #170

sooyeoneo committed:
✅ test : AI 대화형 적성 검사 테스트 시나리오 작성

[테스트 시나리오]
- 정상 답변 플로우 (5가지 적성별)
- 무효 답변 처리 (엉뚱한 답변, 짧은 답변)
- 복합 답변 테스트
- 대화 연속성 확인

[특징]
- 동적 질문 변수 처리 ({{currentQuestion}})
- AI 응답 로깅 및 검증
- 예외 상황 테스트 케이스

[주의사항]
- AI가 매번 다른 질문 생성
- Temperature 0.7로 창의적 대화

Resolves: #64 #67 #98 #151 #170

sooyeoneo committed:
🐛 fix : UserAptitude 엔티티 생성 시 NOT NULL 제약조건 오류 수정

[문제]
- aptitude_code 컬럼 NOT NULL 제약조건 위반
- Builder 패턴 사용 시 aptitudeType 누락

[해결]
- savePartialResult(): 임시 기본값 NATURE 설정
  - `.aptitudeType(AptitudeType.NATURE)`: 임시 기본값 설정
- calculateAndSaveResult(): 계산된 최종 적성값 설정
  - `.aptitudeType(finalAptitude)`: 계산된 최정 적성으로 설정
- selectAptitudeManually(): 수동 선택한 적성값 설정
  - `.aptitudeType(aptitudeType)`: 수동 선택한 적성으로 설정

[영향 범위]
- 적성 검사 시작 및 진행 중 오류 해결
- 부분 저장 기능 정상 동작

Fixes: Column 'aptitude_code' cannot be null 오류

Resolves: #170

sooyeoneo committed:
🐛 fix : 무효 답변 시 aptitudeType null 처리 추가

[문제]
- 엉뚱한 답변 시 aptitudeType이 null로 반환되어 NPE 발생
- updateScoreFromAnalysis에서 toUpperCase() 호출 시 에러

[해결]
- aptitudeType null 체크 추가
- 무효 답변은 점수 업데이트 생략
- 이력 저장 시 "무효 답변" 텍스트 처리

[영향]
- 예외 답변 처리 안정성 향상
- AI 재질문 기능 정상 동작

Fixes: NullPointerException at updateScoreFromAnalysis

Resolves: #170

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

# Conflicts:
#	src/main/java/com/salayo/locallifebackend/global/error/ErrorCode.java

sooyeoneo committed:
🔧 add : NO_INCOMPLETE_TEST 누락된 에러 코드 재추가

sooyeoneo committed:
Merge branch 'develop' into feature/ai-aptitude

# Conflicts:
#	src/main/java/com/salayo/locallifebackend/global/error/ErrorCode.java

sooyeoneo committed:
🔥 remove : AptitudeController에서 @RequiredArgsConstructor 제거 및 생성자 명시

sooyeoneo committed:
♻️ refactor : 클래스 레벨 @PreAuthorize 적용으로 권한 관리 개선

[변경사항]
- 메서드별 @PreAuthorize("hasRole('USER')") 제거
- 클래스 레벨에 @PreAuthorize 일괄 적용

[개선효과]
- 코드 중복 제거 (10개 메서드 → 1개 클래스)
- 권한 누락 방지
- 일관된 권한 관리

[영향범위]
- AptitudeController 전체 메서드 USER 권한 적용
- 기능 변경 없음 (리팩토링만)

Resolves: #55 #170

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/dto/AptitudeResumeResponseDto.java:
```diff
@@ -0,0 +1,26 @@
+	private final Integer totalSteps;
+	private final AptitudeQuestionResponseDto nextQuestion;
+	private final Integer completedSteps;
+	private final String message;
```
`AptitudeResumeResponseDto`의 message 필드가 `CommonResponseDto.message`와 중복되어 프론트에서 혼동될 수도 있을 것 같아요!
혹시 이 필드가 특정 의미가 있다면 그걸로 이름을 교체하는 건 어떨까요?

sooyeoneo replied:
말씀하신 대로 message라는 이름이 CommonResponseDto와 혼동을 줄 수 있겠군요!

해당 필드는 테스트 이어할 때 사용자에게 보여줄 안내 메시지를 담는 역할이므로, 
의도를 더 명확히 하기 위해 `resumeMessage`로 변경했습니다. 

프론트엔드 입장에서 고려하는 습관을 들여야겠네요🥲
API 응답 필드명에 대한 세심한 피드백 감사합니다!

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/entity/AptitudeTestHistory.java:
```diff
@@ -58,11 +58,18 @@ public class AptitudeTestHistory extends BaseEntity {
 	private Boolean isCompleted = false; // 테스트 완료 여부
+	
+	@Column(name = "is_partial_save")
+	private Boolean isPartialSave = false; // 부분 저장 여부
```
`isCompleted, isPartialSave`가 `Boolean`으로 되어 있는데, 이는 Wrapper 타입이라 DB에는 NULL이 들어올 수 있고 엔티티 초기값(= false)과 불일치할 수 있을 것 같아요.

`boolean` 타입 + `nullable = false`로 두는게 더 안전하고 NPE를 예방할 수 있을 것 같아요!!

sooyeoneo replied:
좋은 지적 감사합니다! 🖍️
Boolean 타입이 null을 허용하여 발생할 수 있는 잠재적인 NPE와 데이터 불일치 문제를 짚어주셨네요. 

말씀해주신 대로 boolean 원시 타입과 `@Column(nullable = false)`를 적용하여, 
타입 안정성을 높이고 의도를 명확히 하도록 수정했습니다. 

덕분에 코드가 더 튼튼(?)해졌어요!

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeTestHistoryService.java:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        // 세션은 존재하지만, 권한이 없는 경우
+        if (!histories.getFirst().getMember().getId().equals(memberId)) {
```
제가 로직을 잘못 이해했을 수도 있어요!

`AptitudeTestHistoryRepository`에서 
`List<AptitudeTestHistory> findBySessionIdAndMemberOrderByStepAsc(String sessionId, Member member);`로 member를 조건으로 걸고 있어 다른 멤버의 세션 데이터가 조회되지 않을 것 같은데 한번 더 검증하는 이유가 있을까요?

sooyeoneo replied:
아닙니다! 제가 로직을 잘못 작성한 게 맞습니다 🥹

지난번 리뷰에서 예외 처리를 명확히 하기 위해 소유권 검증 로직을 분리하였는데,
그 이후에 `AptitudeTestHistoryRepository`에서 `findBySessionId()`에 `AndMemeber`를 추가한 점을 놓쳤네요.   

현재의 `findBySessionIdAndMember` 쿼리에서 이미 member 조건으로 필터링하기 때문에, 
불필요한 중복 검증 로직은 삭제하였습니다.

꼼꼼하게 봐주신 덕분에 중복 코드를 제거할 수 있었어요. 감사합니다! 😊

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeTestHistoryService.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    // 미완료 테스트 삭제
+    @Transactional
```
`AptitudeTestHistoryRepository`의 `deleteByMemberAndIsCompletedFalse` 메서드는 네이밍 기반 쿼리지만, 내부적으로 `@Modifying` 쿼리와 같은 동작을 하는 것으로 알고 있는데요, 이 경우 DB에는 즉시 반영되지만 JPA는 트랜잭션 커밋 시점에 flush가 일어나기 때문에 영속성 컨텍스트와 DB의 상태가 어긋날 수 있을 것 같아요.

이 부분은 데이터 정리를 위한 쿼리로 보이는데 전체 트랜잭션과 분리하여 실행하는 것이 안전할 것 같아요!

저도 잘 모르는 부분이라 찾아보니 이런 경우에는 `@Transactional(propagation = REQIRES_NEW)` 를 붙여 개별 트랜잭션으로 동작하도록 하는 것이 좋다고 합니당! 더 좋은 방법이 있다면 알려주세용!

sooyeoneo replied:
벌크 연산 시 발생할 수 있는 영속성 컨텍스트와의 불일치 문제라는 정말 중요한 부분을 짚어주셔서 감사합니다!
제안해주신 `REQUIRES_NEW` 전파 옵션도 좋은 방법이라고 생각합니다. 

저는 영속성 컨텍스트와 DB 상태 불일치 문제에 대해 더 깊이 고민해 보았는데요,
현재 `startTest()` 메서드에서는 삭제 로직이 다른 로직보다 먼저 실행되어 당장 문제가 발생하지는 않습니다. 
처음에는 이 점을 근거로 현상 유지를 생각했습니다.

하지만, 추후에 이 `deleteIncompleteTests()`메서드가 다른 트랜잭션의 중간에서 재사용될 경우를 고려하면, 
메서드 자체에 안정성을 부여하는 것이 더 견고한 설계라는 결론을 내렸습니다. 🧐

> 제안해주신 내용 덕분에 코드의 장기적인 유지보수성까지 생각해보게 되었습니다.

찾아보니, JPA가 제공하는 `@Modifying(clearAutomatically = true)` 옵션을 사용하면 
쿼리 실행 후 영속성 컨텍스트를 자동으로 비워주어 더 직접적으로 문제를 해결할 수 있더군요. 

그래서 저는 `REQUIRES_NEW`(새로운 커넥션을 얻는 등 약간의 오버헤드가 있음)보다 더 가볍고 직접적인 해결책인 
`@Modifying(clearAutomatically = true)` 옵션을 Repository 메서드에 추가하여, 
이 **메서드가 어디서 호출되든 데이터 정합성을 보장**하도록 수정하였습니다.

깊이 있는 리뷰 정말 감사합니다. 덕분에 많이 배웁니다! 👍

j-hann commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeTestHistoryService.java:
```diff
@@ -0,0 +1,107 @@
+    @Transactional(readOnly = true)
+    public List<AptitudeTestHistoryResponseDto> getTestHistoryBySession(Long memberId, String sessionId) {
+        Member member = memberRepository.findById(memberId)
+            .orElseThrow(() -> new CustomException(ErrorCode.MEMBER_NOT_FOUND));
```
`default Member findByIdOrElseThrow(long memberId){
		return findById(memberId).orElseThrow(() -> new CustomException(ErrorCode.MEMBER_NOT_FOUND));
	};`

MemberRepository 쪽에 member id로 조회하는건 자주 사용하는거라 예외처리까지 같이 해주는 메서드 구현해 놓은 것으로 알고 있는데, 혹시 비슷한 메서드를 Service단에서 같은 에러코드로 따로 예외처리 하시는 이유가 있을까요! 
특별한 이유가 없다면 기존 findByIdOrElseThrow 메서드 사용하는 것이 좋을 것 같습니다!

sooyeoneo replied:
지연님, `findByIdOrElseThrow` 헬퍼 메서드를 알려주셔서 감사합니다!

제가 `MemberRepository`를 미처 확인하지 못했네요.
말씀해주신 대로 중복 코드를 줄이고 일관성을 유지하기 위해 기존 메서드를 사용하도록 수정했습니다. 
덕분에 코드가 훨씬 깔끔해졌네요. 😁

j-hann commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeTestHistoryService.java:
```diff
@@ -0,0 +1,107 @@
+    @Transactional(readOnly = true)
+    public List<AptitudeTestHistoryResponseDto> getCompletedTestHistory(Long memberId) {
+        Member member = memberRepository.findById(memberId)
+            .orElseThrow(() -> new CustomException(ErrorCode.MEMBER_NOT_FOUND));
```
이 부분도 특별한 이유가 없다면 기존 findByIdOrElseThrow 메서드 사용하는 것이 좋을 것 같습니다!

sooyeoneo replied:
넵 수정하였습니다!
좋은 제안 감사합니다!👍

j-hann reviewed (approved):
굿굿👍👍

sooyeoneo committed:
♻️ refactor : Boolean 래퍼 타입을 boolean 원시 타입으로 변경

[변경사항]
- isCompleted: Boolean → boolean
- isPartialSave: Boolean → boolean
- nullable = false 제약조건 추가
- 기본값 false 명시

[수정 파일]
- AptitudeTestHistory: 필드 타입 변경
- 관련 Service: getter 메서드명 변경 (getIsCompleted() → isCompleted())

[개선효과]
- NPE 위험 제거
- 불필요한 null 상태 차단
- 데이터 정합성 향상
- 타입 안전성 강화

Resolves: #171

sooyeoneo committed:
🔥 remove : 불필요한 중복 권한 검증 로직 제거

[삭제 내용]
- AptitudeTestHistoryService.getTest()의 중복 권한 검증

[삭제 이유]
- Repository 메서드가 이미 member 조건으로 필터링
- findBySessionIdAndMemberOrderByStepAsc()는 해당 member의 데이터만 반환
- 반환된 결과를 다시 검증하는 것은 논리적으로 불필요

[변경 사항]
- 46-48번 라인 권한 검증 로직 삭제 FORBIDDEN_ACCESS 예외 처리 제거

Resolves: #171

sooyeoneo committed:
🔧 add : 벌크 삭제 시 영속성 컨텍스트 자동 정리 추가

[변경사항]
- deleteByMemberAndIsCompletedFalse()에 @Modifying(clearAutomatically = true) 추가

[변경 이유]
- 현재는 삭제 로직이 트랜잭션 초반에 실행되어 문제없음
- 미래에 다른 트랜잭션 중간에서 재사용될 경우 영속성 컨텍스트 불일치 위험
- 방어적 프로그래밍으로 메서드 자체의 안전성 보장

[기술적 선택]
- REQUIRES_NEW: 새 트랜잭션 생성으로 오버헤드 큼
- clearAutomatically: 현재 영속성 컨텍스트만 정리, 가장 효율적

[개선효과]
- 메서드 호출 위치와 무관하게 데이터 정합성 보장
- 미래의 코드 변경에도 안전
- 암묵적 호출 순서 의존성 제거

Resolves: #171

sooyeoneo committed:
♻️ refactor : MemberRepository 헬퍼 메서드 활용으로 중복 제거

[변경사항]
- findById().orElseThrow() → findByIdOrElseThrow() 사용
- AptitudeService: Member 조회 로직 통일
- AptitudeTestHistoryService: Member 조회 로직 통일

[개선효과]
- 코드 중복 제거 (DRY 원칙 적용)
- Repository 레벨 헬퍼 메서드 활용
- 예외 처리 로직 일원화
- 코드 일관성 향상

[영향 범위]
- 총 8개 메서드의 Member 조회 로직 개선

Resolves: #171

sooyeoneo committed:
♻️ refactor : DTO 필드명을 더 명확하게 변경 (message → resumeMessage)

[변경사항]
- AptitudeResumeResponseDto: message → resumeMessage
- AptitudeService.resumeTest(): 필드명 일관성 적용

[변경 이유]
- CommonResponseDto.message와 혼동 방지
- 필드의 구체적 역할 명시 (테스트 이어하기 안내 메시지)
- DTO 이름(Resume)과 필드명의 의미적 일관성 확보

[개선효과]
- API 응답 구조 명확화
- 프론트엔드 개발 시 혼동 방지
- 코드 가독성 향상

Resolves: #171

sooyeoneo committed:
♻️ refactor : @NotNull을 @Min으로 변경하여 유효성 검증 강화

[변경사항]
- @NotNull → @Min(value = 1)
- 단계 최솟값 검증 추가

[변경 이유]
- @NotNull은 null만 체크, 0이나 음수 허용
- @Min은 null 체크 + 최솟값 검증 동시 수행
- 비즈니스 로직 상 step은 1~5 범위여야 함

[개선효과]
- 유효하지 않은 단계 값(0, 음수) 차단
- 더 엄격한 입력값 검증
- @Schema의 minimum 속성과 일치

Resolves: #171

sooyeoneo committed:
♻️ refactor : 적성 수동 선택 API를 @RequestBody 방식으로 개선

[변경사항]
- @RequestParam → @RequestBody + DTO 방식으로 변경
- AptitudeSelectRequestDto 생성 (record 활용)
- @Valid 어노테이션으로 유효성 검증 추가

[변경 이유]
- 확장성: 추후 필드 추가 시 DTO만 수정하면 됨
- API 명세 명확화: JSON 요청 구조 명시적 정의
- 유효성 검증: @NotNull 등 검증 어노테이션 활용 가능

[기술적 개선]
- Java record로 불변 DTO 구현
- @Valid로 컨트롤러 레벨 검증
- RESTful API 표준 준수

[API 변경]
- Before: POST /ai/aptitude/select?aptitudeType=NATURE
- After: POST /ai/aptitude/select (JSON Body)

Breaking Change: API 요청 방식 변경 (쿼리 파라미터 → Request Body)

Resolves: #171

sooyeoneo committed:
♻️ refactor : AI 적성 검사 HTTP 테스트에 이어하기(Resume) 시나리오 및 체이닝 강화

- 테스트 이어하기(Resume) 시나리오 추가
  - 테스트 중단 후 resume 엔드포인트로 상태 복구 테스트
  - Redis 기반 세션 복구 검증
  - 여러 번 중단/이어하기 시나리오 추가

- 요청 간 변수 연결(Chaining) 강화
  - nextStep, questionText 변수 자동 업데이트
  - 5단계 전체를 연속 실행 가능한 체이닝 테스트
  - 실제 클라이언트 동작과 동일한 흐름 시뮬레이션

- 변수 관리 개선
  - submitAnswer에 필요한 핵심 변수 통일
  - 각 응답에서 자동 변수 업데이트 로직 추가
  - 호환성을 위한 중복 변수 제공

Refs: #151 #164 #166 #170 #171
